### ISSUE TITLE
TypeGuard does not work if it's a callable instances of a class

### ISSUE THREAD

**Issue by flisboac** [microsoft/pyright#4482]
**Describe the bug**
If you implement a type guard as a callable instance of a class, Pyright is unable to detect that the instance itself is indeed a type guard, even if you annotate `__call__` appropriately.

Workaround so far has been to either:
- Call `__call__` explicitly, which is weird; or
- Cast the instance to `Callable[[Any], TypeGuard[T]]` (or equivalent) explicitly.

**To Reproduce**

Follows an example:

```python
from typing import Any

from typing_extensions import TypeGuard


class IsInt:
    def __call__(self, value: Any) -> TypeGuard[int]:
        return isinstance(value, int)


def random() -> Any:
    return 42


value = random()
guard = IsInt()

if guard(value):
    # PYLANCE :: Type of "value" is "Any"
    # MYPY    :: Revealed type is "builtins.int"
    reveal_type(value)
```

I'd expect the same behaviour as mypy.

**Expected behavior**

Pyright should be able to infer that the instance is a type guard itself, and type-guard the `if` block appropriately.

AFAICT, there's nothing in [PEP-647](https://peps.python.org/pep-0647/) that should prohibit such usage. I also don't see why it shouldn't be accepted as a valid usage pattern.

**Screenshots or Code**

_(Already provided previously.)_

Libraries/tools used:

- python == 3.7.9
- typing-extensions == 4.4.0
- mypy == 0.991

**VS Code extension or command-line**

Using the Python extension on VSCode.

- Pylance v2023.1.10

**Comment by erictraut** [comment]
A user-defined type guard function can be a regular function or a method. I'm the author of PEP 647, so I can say with some authority that it is not intended to work with a `__call__` method.

**Comment by erictraut** [comment]
Since mypy chose to implement support for this, I'll change pyright to do the same. I wouldn't recommend counting on this behavior, however, because other type checkers may choose not to support this. As I mentioned, this wasn't intended to work.

**Comment by erictraut** [comment]
This is included in pyright 1.1.290, which I just published. It will also be included in a future release of pylance.

**Comment by flisboac** [comment]
> Since mypy chose to implement support for this, I'll change pyright to do the same. 

That's truly amazing, thank you so much!

> ... other type checkers may choose not to support this

I think there's no issue with that, as long as a team specifies which type checkers they ought to use. Two different implementations should be good enough to cover most cases.

From Python's perspective, a type guard is just a callable returning bool. At runtime, the effect of `guard(value)` and `guard.__call__(value)` are precisely the same.

But I understand, and fully agree with your point.

> I can say with some authority that it is not intended to work with a __call__ method.

Any chance this behaviour could be standardized (or at least clarified) in e.g. a future PEP?

**Comment by erictraut** [comment]
If it's internal to your code base, that should be fine. I wouldn't recommend using it in the public interface of a library.
